📝 Clarify misleading "Required, can be `None`" section in `query-params-str-validations.md`

ChaitanyaSai-Meka opened the pull request:
The previous explanation suggested that a query parameter could be both required and accept a true `None` value.
However, since HTTP query parameters are transmitted as strings, clients cannot actually send `None` or `null`.

This update clarifies that behavior and adds an example showing how to manually handle `"None"` or empty strings as `None` inside a route.

ChaitanyaSai-Meka committed:
docs: clarify misleading 'Required, can be None' section in query param validation

YuriiMotov reviewed (changes requested):
@ChaitanyaSai-Meka, thanks for your interest and efforts!

Personally, I like the explanation!

We need to fix formatting (format `None`, `q=None`, etc.. with back-ticks), move code example to separate `.py` file in  `docs_src/query_params_str_validations/`, and cover it by test (tests for docs examples are in `tests/test_tutorial/query_params_str_validations/`)

ChaitanyaSai-Meka committed:
docs: fix inline code formatting and add tested example for Required, can be None section

pre-commit-ci[bot] committed:
🎨 [pre-commit.ci] Auto format from pre-commit.com hooks

ChaitanyaSai-Meka committed:
docs: finalize Required, can be None example with test verification

YuriiMotov reviewed (commented):
Thanks!

A few more suggestions!)

YuriiMotov commented on docs_src/query_params_str_validations/tutorial006d_an_py310.py:
```diff
@@ -0,0 +1,10 @@
+app = FastAPI()
+
+@app.get("/items/")
+async def read_items(q: Annotated[str | None, Query()] = ...):
```
I can help with the last part (adding code example with `BeforeValidator`) if you need

ChaitanyaSai-Meka committed:
style: clean formatting with Ruff + Black

pre-commit-ci[bot] committed:
🎨 [pre-commit.ci] Auto format from pre-commit.com hooks

ChaitanyaSai-Meka commented:
@YuriiMotov
Implemented all requested changes:

Removed ellipsis and kept min_length=3 consistent

Added BeforeValidator example for Pydantic v2

Expanded test coverage for ("None", "", "null", "hello")

Updated “request validation error” wording

Verified locally 
Ready for re-review

ChaitanyaSai-Meka commented:
Hi @YuriiMotov 👋

I’ve confirmed that all requested changes have been implemented, 
rebased the branch on the latest `master`, and verified that local tests pass successfully ✅

It seems the CI jobs failed due to cached environment setup, not due to the changes themselves.  
Could you please re-run the workflows when convenient?

Thank you for reviewing and helping out!

YuriiMotov commented on docs_src/query_params_str_validations/tutorial006d_an_py310.py:
```diff
@@ -0,0 +1,19 @@
+from typing import Annotated, Optional, Union
```
```suggestion
from typing import Optional, Union
from typing_extensions import Annotated
```

Otherwise it fails for Python 3.8

ChaitanyaSai-Meka replied:
Thanks @YuriiMotov!
I’ve updated th import to use typing_extensions.Annotated for Python 3.8 compatibility ✅
Just pushed the fix , everything should be ready now!

ChaitanyaSai-Meka committed:
fix: import Annotated from typing_extensions for Python 3.8 compatibility

pre-commit-ci[bot] committed:
🎨 [pre-commit.ci] Auto format from pre-commit.com hooks

ChaitanyaSai-Meka committed:
test: finalize client fixture and absolute import for needs_pydanticv2

pre-commit-ci[bot] committed:
🎨 [pre-commit.ci] Auto format from pre-commit.com hooks

ChaitanyaSai-Meka commented on tests/test_tutorial/test_query_params_str_validations/test_tutorial006d_an_py310.py:
```diff
@@ -0,0 +1,21 @@
+
+from docs_src.query_params_str_validations.tutorial006d_an_py310 import app
+
+client = TestClient(app)
```
Thanks @YuriiMotov!
I’ve rebased the branch and pushed the final test update (tests.utils import).
Verified locally, all works perfectly ✅

ChaitanyaSai-Meka committed:
docs: clarify BeforeValidator is available only in Pydantic v2

YuriiMotov commented on docs/en/docs/tutorial/query-params-str-validations.md:
```diff
@@ -260,11 +260,25 @@ So, when you need to declare a value as required while using `Query`, you can si
+
+If you want to accept special values (like `"None"` or an empty string) and interpret them as `None` in your application, you can handle them manually in your function:
+
+{* ../../docs_src/query_params_str_validations/tutorial006d_an_py310.py hl[9,11] *}
```
```suggestion
{* ../../docs_src/query_params_str_validations/tutorial006d_an_py310.py hl[9:13 18] *}
```

YuriiMotov reviewed (approved):
LGTM!

@ChaitanyaSai-Meka, thanks!

YuriiMotov committed:
Apply suggestions from code review

YuriiMotov committed:
Update docs/en/docs/tutorial/query-params-str-validations.md

YuriiMotov committed:
Update highlights

ChaitanyaSai-Meka commented:
Hi @YuriiMotov 👋
Just wanted to check if this PR is now queued for merge. No rush at all , just noticed that all checks have passed and it's been a few weeks, so I thought I'd follow up. Thanks again for the guidance!

YuriiMotov commented:
> Hi @YuriiMotov 👋 Just wanted to check if this PR is now queued for merge. No rush at all , just noticed that all checks have passed and it's been a few weeks, so I thought I'd follow up. Thanks again for the guidance!

Hi!
Yes, we are waiting for final review by Sebastian

YuriiMotov committed:
Rename code example file and add variants

pre-commit-ci-lite[bot] committed:
🎨 Auto format

YuriiMotov committed:
Update text in docs

YuriiMotov committed:
Merge branch 'master' into docs/fix-required-can-be-none

YuriiMotov committed:
Remove py39 code example

YuriiMotov committed:
Merge remote-tracking branch 'upstream/master' into docs/fix-required-can-be-none

YuriiMotov committed:
Merge branch 'master' into docs/fix-required-can-be-none

Shaan-alpha commented:
Thanks for clarifying and for continuing the work on this PR.

I had independently worked on the same docs behavior around required nullable query parameters in #15514, so I’ll follow this thread instead.

If any additional cleanup, tests, wording improvements, or review support would still be useful here, I’d be happy to help contribute.
